File: August/Week5/Day3/deleteNodeBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = function (root, key) {
    if (root === null) {
        return root;
    } else if (key < root.val) {
        root.left = deleteNode(root.left, key);
    } else if (key > root.val) {
        root.right = deleteNode(root.right, key);
    } else {
        if (root.left === null && root.right === null) {
            root = null;
        } else if (root.left === null && root.right !== null) {
            root = root.right;
        } else if (root.left !== null && root.right === null) {
            root = root.left;
        } else {
            let minVal = findMinNode(root.right);
            root.val = minVal.val;
            root.right = deleteNode(root.right, root.val);
        }
    }

    return root;
};

var findMinNode = function (root) {
    let current = root;
    while (current.left !== null) {
        current = current.left;
    }

    return current;
};

File: August/Week5/Day3/deleteNodeBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        # print(root,key)
        if root == None:
            return root
        
        if key < root.val:
            root.left = self.deleteNode(root.left,key)
        elif key > root.val:
            root.right = self.deleteNode(root.right,key)
            
        else:
            if root.left == None and root.right == None:
                root = None
            
            elif root.left == None and root.right != None:
                root = root.right
            elif root.left != None and root.right == None:
                root = root.left
            else:
                minVal = self.findMinNode(root.right)
                root.val = minVal.val
                root.right = self.deleteNode(root.right, root.val)
                
        return root
                
    def findMinNode(self, root: TreeNode) -> TreeNode:
        current = root
        while(current.left is not None):
            current = current.left
            
        return current
            